### PR TITLE
{Package} Migrate CLI packaging system to PEP420

### PR DESCRIPTION
**Pull request by arrownj**
**Description**  

Background
----------
[PEP420 ](https://www.python.org/dev/peps/pep-0420/)is an official python enhancement proposal which introduces implicit namespace package. 

There are 3 ways to declare a namespace package.

- pkg_resource style
- pkgutil style
- native style (PEP420)

Currently, azure cli is using pkg_resource style namespace package, which is no longer recommended per python official document. Besides, pkg_resource style namespace package performs slow when importing python modules.

This PR is to migrate azure cli namespace package to native style. Refer [Packaging namespace packages ](https://packaging.python.org/guides/packaging-namespace-packages/)to see details about above 3 namespace package styles.

There is also a guide about how to do azure sdk package. Refer [Azure packaging ](https://github.com/Azure/azure-sdk-for-python/blob/master/doc/dev/packaging.md)to see the details. 

We will follow this guide to to do azure cli packaging.

Goal
----------
The main goal of this task to to remove `__init__.py` file under `azure` and `cli` folder in our final distributions on different platforms, includes:

- MSI
- DEB
- RPM
- Pypi
- Homebrew

Changes
----------
In order to achieve this, several changes are needed as below:

**azure-cli-dev-tools**
- remove installation of all cli related `-nspkg` packages([PR](https://github.com/Azure/azure-cli-dev-tools/pull/240))

**some python sdk dependency**
- azure-cosmos
- azure-multiapi-storage

**azure-cli**
- remove all `-nspkg` dependency
- remove `azure-cli-nspkg` and `azure-cli-command_modules-nspkg`
- update `__init__.py`, `setup.py` and `MANIFEST.in` according to azure packaging guide
- move `__version__` and `__author__` to `__main__.py`
- update `azure_bdist_wheel` to fix package `__main__.py` problem
- sdk dependency
- fix test cases

Plan
----------

1. merge `azure-cli-dev-tools` [PR](https://github.com/Azure/azure-cli-dev-tools/pull/240)
2. build new `azdev` version
3. rollback `azdev` updates in this PR
4. merge PR

**Testing Guide**  
- Install `azure-cli` on different platform, and see whether `__init__.py` still exists
- Make sure all previous tests are passed in both current CI pipelines and Feng's new CI pipelines.
- Make sure `.tar.gz` and `.whl` files are same with previous ones.

**History Notes**  

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.  
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by arrownj** [comment]
PEP420

File: src/azure-cli/azure/cli/command_modules/eventgrid/linter_exclusions.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+---
+eventgrid partner topic event-subscription update:
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/eventgrid/linter_exclusions.yml:2]: This should be in global suppression already.

File: scripts/ci/install_azdev.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
```
Comment (MyronFanQiu) [inline scripts/ci/install_azdev.sh:1]: This file should be deleted as well.

Comment (jiasli) [reply scripts/ci/install_azdev.sh:1]: Looks like this file is not referenced anywhere.

File: azure-pipelines.yml [hunk 3]
```diff
@@ -485,7 +485,7 @@ jobs:
   displayName: Build Homebrew Formula
 
   dependsOn: BuildPythonWheel
-  condition: and(succeeded(), in(variables['Build.Reason'], 'IndividualCI', 'BatchedCI', 'Manual'))
```
Comment (haroldrandom) [inline azure-pipelines.yml:488]: No need to build at those trigger condition?

Comment (arrownj) [reply azure-pipelines.yml:488]: yes, will revert these changes before merge it. Just test whether these pipelines can pass.

File: .snapcraft.yaml [hunk 4]
```diff
@@ -28,7 +28,7 @@ parts:
       - libssl1.0.0
     requirements: requirements.txt
     install: |
-      python_packages="src/azure-cli-core src/azure-cli-nspkg src/azure-cli-command_modules-nspkg src/command_modules/azure-cli*"
+      python_packages="src/azure-cli-core src/command_modules/azure-cli*"
```
Comment (fengzhou-msft) [inline .snapcraft.yaml:31]: Is this for building snap package?

File: src/azure-cli/azure/cli/command_modules/config/tests/latest/test_config.py [hunk 5]
```diff
@@ -13,6 +13,7 @@
 
 class ConfigTest(ScenarioTest):
 
+    @unittest.skip('skip')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/config/tests/latest/test_config.py]: Maybe you can replace 'skip' with a real reason.

Comment (arrownj) [reply src/azure-cli/azure/cli/command_modules/config/tests/latest/test_config.py]: It's already outdated.

**Review by fengzhou-msft — APPROVED** [review]
LGTM, please test the upgrade case as well as the new install case.

**Review by haroldrandom — APPROVED** [review]

**Comment by zhoxing-ms** [comment]
Just be curiously, our relase pipeline also publishes pakage to the following platforms:
`ACR`, `Windows Artifacts`, `Cloud Shell`

Will these changes affect the packages for the above platforms?
If so, have they been tested and evaluated?

**Comment by arrownj** [comment]
> Cloud



> Just be curiously, our relase pipeline also publishes pakage to the following platforms:
> `ACR`, `Windows Artifacts`, `Cloud Shell`
> 
> Will these changes affect the packages for the above platforms?
> If so, have they been tested and evaluated?

We don't consider the scenario like that. We build package on Windows, MAC, pypi, RPM, DEB, the 3 target platforms are all using these distributions. We have test it on all OS platforms. So I think it can cover the scenario you mentioned.

**Comment by zhoxing-ms** [comment]
> We don't consider the scenario like that. We build package on Windows, MAC, pypi, RPM, DEB, the 3 target platforms you mentioned are all using these distributions. We have test it on all OS platforms. So I think it can cover the scenario you mentioned.

Got it~ thanks

**Review by zhoxing-ms — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/config/tests/latest/test_config.py [hunk 6]
```diff
@@ -24,6 +24,7 @@ def test_config(self):
 
         # C:\Users\{username}\AppData\Local\Temp
         tempdir = tempfile.gettempdir()
+        tempdir = os.path.realpath(tempdir)
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/config/tests/latest/test_config.py:27]: Let's add a comment for this special case on MacOS.

**Review by jiasli — APPROVED** [review]
My parts LGTM.

File: src/azure-cli/azure/cli/command_modules/config/tests/latest/test_config.py (hunk 6 above)
Comment (arrownj) [reply src/azure-cli/azure/cli/command_modules/config/tests/latest/test_config.py]: updated.

**Comment by jiasli** [comment]
This PR was first included in Azure CLI 2.11.0.

**Pull request merged**